{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8a7c36a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\antoi\\Documents\\Fichier Antoine\\DTU\\Thesis\\thesis_env\\Lib\\site-packages\\xenaPython\\__init__.py:109: FutureWarning: Possible nested set at position 7\n",
      "  re.sub(r\"^[^[]+[[]([^]]*)[]].*$\", r\"\\1\", query, flags=re.DOTALL))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing HER2...\n",
      "Fetching transcripts for gene with Ensembl ID: ENSG00000141736...\n",
      "Fetching protein sequence for Ensembl IDs: ENST00000584601 ENST00000584014 ENST00000578199 ENST00000445658 ENST00000584450 ENST00000269571 ENST00000578502 ENST00000578709 ENST00000582818 ENST00000580074 ENST00000863095 ENST00000863096 ENST00000863097 ENST00000863098 ENST00000863099 ENST00000863100 ENST00000863101 ENST00000863102 ENST00000863103 ENST00000938923 ENST00000938924 ENST00000938925 ENST00000959774 ENST00000959775\n",
      "Saved transcript_to_isoform_mapping.csv to ./files/HER2\n",
      "Saved sequences_data.json to ./files/HER2\n",
      "Processing CD20...\n",
      "Fetching transcripts for gene with Ensembl ID: ENSG00000156738...\n",
      "Fetching protein sequence for Ensembl IDs: ENST00000532491 ENST00000532073 ENST00000345732 ENST00000534668 ENST00000528313 ENST00000533306 ENST00000674194 ENST00000389939 ENST00000904593 ENST00000904594 ENST00000966396\n",
      "Saved transcript_to_isoform_mapping.csv to ./files/CD20\n",
      "Saved sequences_data.json to ./files/CD20\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from topology import (\n",
    "    ensure_dir,\n",
    "    get_transcripts_and_sequences,\n",
    "    align_protein_sequences,\n",
    "    run_deeptmhmm,\n",
    "    generate_isoform_mapping,\n",
    "    create_membrane_topology_objects\n",
    ")\n",
    "from expression import (\n",
    "    get_expression_data,\n",
    "    plot_expression\n",
    ")\n",
    "import json\n",
    "\n",
    "## Variables\n",
    "\n",
    "# Output directory\n",
    "out_dir = \"./files\"\n",
    "out_dir_for_plots = \"./app/files_for_plots\"\n",
    "\n",
    "# Proteins we want to show\n",
    "proteins = [\"HER2\", \"CD20\"]\n",
    "\n",
    "# Email for alignment\n",
    "email = \"s242830@dtu.dk\"\n",
    "\n",
    "# Variables for each protein\n",
    "proteins_ids = {\n",
    "    \"HER2\": \"ENSG00000141736\",\n",
    "    \"CD20\": \"ENSG00000156738\"\n",
    "}\n",
    "\n",
    "# Running the analysis for each protein\n",
    "for protein in proteins:\n",
    "    print(f\"Processing {protein}...\")\n",
    "    ensure_dir(out_dir + \"/\" + protein)\n",
    "    get_transcripts_and_sequences(proteins_ids[protein], out_dir + \"/\" + protein)\n",
    "    align_protein_sequences(email, out_dir + \"/\" + protein)\n",
    "    run_deeptmhmm(out_dir + \"/\" + protein)\n",
    "    mapping = generate_isoform_mapping(out_dir + \"/\" + protein)\n",
    "    json.dump(mapping, open(out_dir + \"/\" + protein + \"/transcript_to_isoform_mapping.json\", \"w\"))\n",
    "    create_membrane_topology_objects(mapping, out_dir + \"/\" + protein)\n",
    "    \n",
    "\n",
    "    # with open(out_dir + \"/HER2/isoforms.fasta\") as f: \n",
    "    #     lines = f.readlines()\n",
    "    #     headers = [line.strip()[1:] for line in lines if line.startswith(\">\")]\n",
    "    #     all_ids = []\n",
    "    #     for h in headers:\n",
    "    #         all_ids.extend(h.split(\"|\"))\n",
    "\n",
    "    # get_expression_data(all_ids, out_dir + \"/HER2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "682e6493",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pushing relevant files to the app for plotting\n",
    "\n",
    "for protein in proteins:\n",
    "    ensure_dir(out_dir_for_plots + \"/\" + protein)\n",
    "    os.system(f\"cp {out_dir}/{protein}/transcript_to_isoform_mapping.json {out_dir_for_plots}/{protein}/transcript_to_isoform_mapping.json\")\n",
    "    os.system(f\"cp {out_dir}/{protein}/membrane_topology.csv {out_dir_for_plots}/{protein}/membrane_topology.csv\")\n",
    "    os.system(f\"cp {out_dir}/{protein}/sequences_data.json {out_dir_for_plots}/{protein}/sequences_data.json\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
