{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8a7c36a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing PROM1...\n",
      "Fetching transcripts for gene with Ensembl ID: ENSG00000007062...\n",
      "Fetching protein sequence for Ensembl IDs: ENST00000447510 ENST00000505450 ENST00000508167 ENST00000508322 ENST00000508940 ENST00000510224 ENST00000514967 ENST00000539194 ENST00000540805 ENST00000675377 ENST00000675613 ENST00000888869 ENST00000888870 ENST00000888871 ENST00000888872 ENST00000888873 ENST00000888874 ENST00000888875 ENST00000888876 ENST00000888877 ENST00000888878 ENST00000888879 ENST00000888880 ENST00000888881 ENST00000938908 ENST00000938909 ENST00000938910 ENST00000938911 ENST00000938912 ENST00000971409 ENST00000971410 ENST00000971411\n",
      "ENST00000447510\n",
      "ENST00000505450\n",
      "ENST00000508322\n",
      "ENST00000508940\n",
      "ENST00000514967\n",
      "ENST00000539194\n",
      "ENST00000540805\n",
      "ENST00000675377\n",
      "ENST00000675613\n",
      "ENST00000888869\n",
      "ENST00000888870\n",
      "ENST00000888871\n",
      "ENST00000888872\n",
      "ENST00000938908\n",
      "ENST00000938909\n",
      "ENST00000938910\n",
      "ENST00000971409\n",
      "Submitting alignment job...\n",
      "Successfully submitted. Job ID: clustalo-R20260128-124726-0871-55036046-p1m\n",
      "Job status: RUNNING\n",
      "Job status: FINISHED\n",
      "Fetching alignment result...\n",
      "2026-01-28 13:47:40,343 | INFO : Loaded project DTU/DeepTMHMM:1.0.50\n",
      "Running DeepTMHMM...\n",
      "2026-01-28 13:47:40,991 | INFO : View the result in your browser at: https://biolib.com/results/02dbac4d-0c3a-4062-ba90-e85a72d2bbae/\n",
      "2026-01-28 13:47:42,414 | INFO : Cloud: The job has been queued. Please wait...\n",
      "2026-01-28 13:47:45,020 | INFO : Cloud: Initializing\n",
      "2026-01-28 13:47:47,194 | INFO : Cloud: Pulling images...\n",
      "2026-01-28 13:47:47,194 | INFO : Cloud: Computing...\n",
      "2026-01-28 12:47:48,246 | INFO : Large input detected. Allocating dedicated capacity ...\n",
      "Running DeepTMHMM on 17 sequences...\n",
      "Step 1/4 | Loading transformer model...\n",
      "\n",
      "Step 2/4 | Generating embeddings for sequences...\n",
      "Generating embeddings: 100% 17/17 [00:05<00:00,  2.90seq/s]\n",
      "\n",
      "Step 3/4 | Predicting topologies for sequences in batches of 1...\n",
      "Topology prediction: 100% 17/17 [00:21<00:00,  1.29s/seq]\n",
      "\n",
      "Step 4/4 | Generating output...\n",
      "2026-01-28 13:56:08,972 | INFO : Cloud: Computation finished\n",
      "2026-01-28 12:56:08,367 | INFO : Done in 500.54 seconds\n",
      "2026-01-28 13:56:11,129 | INFO : Cloud: Result Ready\n",
      "2026-01-28 13:56:11,133 | INFO : Waiting for job 02dbac4d-0c3a-4062-ba90-e85a72d2bbae to finish...\n",
      "2026-01-28 13:56:11,572 | INFO : Result 02dbac4d-0c3a-4062-ba90-e85a72d2bbae has finished.\n",
      "2026-01-28 13:56:12,213 | INFO : Saving 3 files to ./files/PROM1/DeepTMHMM_results/...\n",
      "DeepTMHMM run completed.\n",
      "Saved sequences_data.json to ./files/PROM1\n",
      "\u001b[31mWarning: Transcript ENST00000675377 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000675613 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888869 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888870 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888871 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888872 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888873 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888874 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888875 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888876 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888877 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888878 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888879 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888880 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000888881 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938908 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938909 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938910 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938911 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938912 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000971409 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000971410 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000971411 not found in the database.\u001b[0m\n",
      "Creating expression figure objects...\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from scripts.topology import (\n",
    "    ensure_dir,\n",
    "    get_transcripts_and_sequences,\n",
    "    align_protein_sequences,\n",
    "    run_deeptmhmm,\n",
    "    create_membrane_topology_objects\n",
    ")\n",
    "from scripts.expression import (\n",
    "    getting_expression_data,\n",
    "    create_expression_figure_objects\n",
    ")\n",
    "import json\n",
    "import shelve\n",
    "\n",
    "## Variables\n",
    "\n",
    "# Output directory\n",
    "out_dir = \"./files\"\n",
    "out_dir_for_plots = \"./app/files_for_plots\"\n",
    "\n",
    "# Proteins we want to show\n",
    "proteins = [\"HER2\", \"CD20\", \"PROM1\"]\n",
    "\n",
    "# Email for alignment\n",
    "email = \"s242830@dtu.dk\"\n",
    "\n",
    "# Variables for each protein\n",
    "proteins_ids = {\n",
    "    \"HER2\": \"ENSG00000141736\",\n",
    "    \"CD20\": \"ENSG00000156738\",\n",
    "    \"PROM1\": \"ENSG00000007062\"\n",
    "}\n",
    "\n",
    "# Running the analysis for each protein\n",
    "for protein in proteins:\n",
    "\n",
    "    print(f\"Processing {protein}...\")\n",
    "    ensure_dir(out_dir)\n",
    "    transcripts_id, mapping = get_transcripts_and_sequences(proteins_ids[protein], out_dir)\n",
    "    align_protein_sequences(email, out_dir)\n",
    "    run_deeptmhmm(out_dir)    \n",
    "\n",
    "    with shelve.open(out_dir_for_plots + \"/transcripts_to_isoforms_mapping\") as db:\n",
    "        db[protein] = mapping\n",
    "\n",
    "    membrane_topology_object = create_membrane_topology_objects(mapping, out_dir)\n",
    "\n",
    "    with shelve.open(out_dir_for_plots + \"/membrane_topology_objects\") as db:\n",
    "        db[protein] = membrane_topology_object\n",
    "    \n",
    "    expression_df = getting_expression_data(transcripts_id, out_dir)\n",
    "    TCGA_GTEx_plotting_data = create_expression_figure_objects(out_dir, expression_df, mapping)\n",
    "    \n",
    "    with shelve.open(out_dir_for_plots + \"/TCGA_GTEx_plotting_data\") as db:\n",
    "        db[protein] = TCGA_GTEx_plotting_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
