{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b8a7c36a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\antoi\\Documents\\Fichier Antoine\\DTU\\Thesis\\thesis_env\\Lib\\site-packages\\xenaPython\\__init__.py:109: FutureWarning: Possible nested set at position 7\n",
      "  re.sub(r\"^[^[]+[[]([^]]*)[]].*$\", r\"\\1\", query, flags=re.DOTALL))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing HER2...\n",
      "Fetching transcripts for gene with Ensembl ID: ENSG00000141736...\n",
      "Fetching protein sequence for Ensembl IDs: ENST00000584601 ENST00000584014 ENST00000578199 ENST00000445658 ENST00000584450 ENST00000269571 ENST00000578502 ENST00000578709 ENST00000582818 ENST00000580074 ENST00000863095 ENST00000863096 ENST00000863097 ENST00000863098 ENST00000863099 ENST00000863100 ENST00000863101 ENST00000863102 ENST00000863103 ENST00000938923 ENST00000938924 ENST00000938925 ENST00000959774 ENST00000959775\n",
      "Submitting alignment job...\n",
      "Successfully submitted. Job ID: clustalo-R20251219-123419-0287-79626582-p1m\n",
      "Job status: QUEUED\n",
      "Job status: FINISHED\n",
      "Fetching alignment result...\n",
      "2025-12-19 13:34:32,436 | INFO : Loaded project DTU/DeepTMHMM:1.0.44\n",
      "Running DeepTMHMM...\n",
      "2025-12-19 13:34:33,103 | INFO : View the result in your browser at: https://biolib.com/results/9dcd8e47-9a3c-4f8c-a86d-4c0c033d0267/\n",
      "2025-12-19 13:34:34,536 | INFO : Cloud: The job has been queued. Please wait...\n",
      "2025-12-19 13:34:36,963 | INFO : Cloud: The job has been queued. Please wait...\n",
      "2025-12-19 13:34:40,557 | INFO : Cloud: Initializing\n",
      "2025-12-19 13:34:42,710 | INFO : Cloud: Pulling images...\n",
      "2025-12-19 13:34:42,711 | INFO : Cloud: Computing...\n",
      "2025-12-19 12:34:43,030 | INFO : Large input detected. Allocating dedicated capacity ...\n",
      "Running DeepTMHMM on 23 sequences...\n",
      "Step 1/4 | Loading transformer model...\n",
      "\n",
      "Step 2/4 | Generating embeddings for sequences...\n",
      "Generating embeddings: 100% 23/23 [00:12<00:00,  1.87seq/s]\n",
      "\n",
      "Step 3/4 | Predicting topologies for sequences in batches of 1...\n",
      "Topology prediction: 100% 23/23 [01:01<00:00,  2.69s/seq]\n",
      "\n",
      "Step 4/4 | Generating output...\n",
      "2025-12-19 12:38:55,481 | INFO : Done in 252.78 seconds\n",
      "2025-12-19 13:38:57,350 | INFO : Cloud: Computation finished\n",
      "2025-12-19 13:38:57,350 | INFO : Cloud: Result Ready\n",
      "2025-12-19 13:38:57,350 | INFO : Waiting for job 9dcd8e47-9a3c-4f8c-a86d-4c0c033d0267 to finish...\n",
      "2025-12-19 13:38:57,725 | INFO : Result 9dcd8e47-9a3c-4f8c-a86d-4c0c033d0267 has finished.\n",
      "2025-12-19 13:38:58,461 | INFO : Saving 3 files to ./files/HER2/DeepTMHMM_results/...\n",
      "DeepTMHMM run completed.\n",
      "Saved sequences_data.json to ./files/HER2\n",
      "\u001b[31mWarning: Transcript ENST00000863095 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863096 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863097 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863098 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863099 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863100 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863101 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863102 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000863103 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938923 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938924 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000938925 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000959774 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000959775 not found in the database.\u001b[0m\n",
      "Creating expression figure objects...\n",
      "Processing CD20...\n",
      "Fetching transcripts for gene with Ensembl ID: ENSG00000156738...\n",
      "Fetching protein sequence for Ensembl IDs: ENST00000532491 ENST00000532073 ENST00000345732 ENST00000534668 ENST00000528313 ENST00000533306 ENST00000674194 ENST00000389939 ENST00000904593 ENST00000904594 ENST00000966396\n",
      "Submitting alignment job...\n",
      "Successfully submitted. Job ID: clustalo-R20251219-123916-0572-10916693-p1m\n",
      "Job status: QUEUED\n",
      "Job status: FINISHED\n",
      "Fetching alignment result...\n",
      "2025-12-19 13:39:29,636 | INFO : Loaded project DTU/DeepTMHMM:1.0.44\n",
      "Running DeepTMHMM...\n",
      "2025-12-19 13:39:30,766 | INFO : View the result in your browser at: https://biolib.com/results/5eb415e1-bc40-4a23-9e35-5dc7c8115284/\n",
      "2025-12-19 13:39:32,227 | INFO : Cloud: The job has been queued. Please wait...\n",
      "2025-12-19 13:39:34,869 | INFO : Cloud: Initializing\n",
      "2025-12-19 13:39:36,986 | INFO : Cloud: Pulling images...\n",
      "2025-12-19 13:39:36,990 | INFO : Cloud: Computing...\n",
      "2025-12-19 12:39:42,914 | INFO : Extracted zip file to: output/\n",
      "2025-12-19 12:39:42,914 | INFO : Done in 4.98 seconds\n",
      "2025-12-19 13:39:43,485 | INFO : Cloud: Computation finished\n",
      "2025-12-19 13:39:43,485 | INFO : Cloud: Result Ready\n",
      "2025-12-19 13:39:43,487 | INFO : Waiting for job 5eb415e1-bc40-4a23-9e35-5dc7c8115284 to finish...\n",
      "2025-12-19 13:39:43,947 | INFO : Result 5eb415e1-bc40-4a23-9e35-5dc7c8115284 has finished.\n",
      "2025-12-19 13:39:44,732 | INFO : Saving 3 files to ./files/CD20/DeepTMHMM_results/...\n",
      "DeepTMHMM run completed.\n",
      "Saved sequences_data.json to ./files/CD20\n",
      "\u001b[31mWarning: Transcript ENST00000674194 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000904593 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000904594 not found in the database.\u001b[0m\n",
      "\u001b[31mWarning: Transcript ENST00000966396 not found in the database.\u001b[0m\n",
      "Creating expression figure objects...\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from scripts.cellular_localization import run_deeploc2\n",
    "from scripts.topology import (\n",
    "    ensure_dir,\n",
    "    get_transcripts_and_sequences,\n",
    "    align_protein_sequences,\n",
    "    run_deeptmhmm,\n",
    "    generate_isoform_mapping,\n",
    "    create_membrane_topology_objects\n",
    ")\n",
    "from scripts.expression import (\n",
    "    getting_expression_data,\n",
    "    create_expression_figure_objects\n",
    ")\n",
    "import json\n",
    "import shelve\n",
    "\n",
    "## Variables\n",
    "\n",
    "# Output directory\n",
    "out_dir = \"./files\"\n",
    "out_dir_for_plots = \"./app_v2/files_for_plots\"\n",
    "\n",
    "# Proteins we want to show\n",
    "proteins = [\"HER2\", \"CD20\"]\n",
    "\n",
    "# Email for alignment\n",
    "email = \"s242830@dtu.dk\"\n",
    "\n",
    "# Variables for each protein\n",
    "proteins_ids = {\n",
    "    \"HER2\": \"ENSG00000141736\",\n",
    "    \"CD20\": \"ENSG00000156738\"\n",
    "}\n",
    "\n",
    "# Running the analysis for each protein\n",
    "for protein in proteins:\n",
    "\n",
    "    print(f\"Processing {protein}...\")\n",
    "    ensure_dir(out_dir + \"/\" + protein)\n",
    "    transcripts_id = get_transcripts_and_sequences(proteins_ids[protein], out_dir + \"/\" + protein)\n",
    "    align_protein_sequences(email, out_dir + \"/\" + protein)\n",
    "    run_deeptmhmm(out_dir + \"/\" + protein)\n",
    "    mapping = generate_isoform_mapping(out_dir + \"/\" + protein)\n",
    "\n",
    "    with shelve.open(out_dir_for_plots + \"/transcripts_to_isoforms_mapping\") as db:\n",
    "        db[protein] = mapping\n",
    "\n",
    "    membrane_topology_object = create_membrane_topology_objects(mapping, out_dir + \"/\" + protein)\n",
    "\n",
    "    with shelve.open(out_dir_for_plots + \"/membrane_topology_objects\") as db:\n",
    "        db[protein] = membrane_topology_object\n",
    "    \n",
    "    expression_df = getting_expression_data(transcripts_id, out_dir + \"/\" + protein)\n",
    "    TCGA_GTEx_plotting_data = create_expression_figure_objects(out_dir + \"/\" + protein, expression_df, mapping)\n",
    "    \n",
    "    with shelve.open(out_dir_for_plots + \"/TCGA_GTEx_plotting_data\") as db:\n",
    "        db[protein] = TCGA_GTEx_plotting_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
